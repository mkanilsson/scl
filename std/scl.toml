[package]
name = "std"
description = "Scl's standard library"
type = "lib"
