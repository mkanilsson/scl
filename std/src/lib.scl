extern proc exit(i32) !
