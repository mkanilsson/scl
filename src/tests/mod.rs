use std::{env, fs, path::PathBuf, process::Command};

use insta::glob;
use miette::{GraphicalReportHandler, GraphicalTheme};
use nanoid::nanoid;

use crate::{codegen::Codegen, package::Package, typechecker::Checker};

#[test]
fn runtime() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    fn generate_tempfile(extension: &str) -> PathBuf {
        let name = format!("{}.{extension}", nanoid!());
        env::temp_dir().join(name)
    }

    glob!("sources/runtime/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let checked_main_package = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .unwrap();

        insta::assert_debug_snapshot!(checked_main_package);

        let all_units = checked_std_package
            .units
            .into_iter()
            .chain(checked_main_package.units.into_iter())
            .collect::<Vec<_>>();

        let mut codegener = Codegen::new(all_units, checker);
        let code = codegener.generate();

        insta::assert_snapshot!(code);

        let tmp = generate_tempfile("ssa");
        let asm = generate_tempfile("S");
        let executable = generate_tempfile("out");

        std::fs::write(&tmp, code).unwrap();

        let qbe_cmd = Command::new("qbe")
            .arg("-o")
            .arg(&asm)
            .arg(tmp)
            .output()
            .unwrap();

        if !qbe_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&qbe_cmd.stdout);
            let stderr = String::from_utf8_lossy(&qbe_cmd.stderr);

            panic!("QBE failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let gcc_cmd = Command::new("gcc")
            .arg("-o")
            .arg(&executable)
            .arg(&asm)
            .output()
            .unwrap();

        if !gcc_cmd.status.success() {
            let stdout = String::from_utf8_lossy(&gcc_cmd.stdout);
            let stderr = String::from_utf8_lossy(&gcc_cmd.stderr);

            panic!("gcc failed:\nOut:\n{stdout}Err:\n{stderr}");
        }

        let run_cmd = Command::new(executable).output().unwrap();

        let stdout = String::from_utf8_lossy(&run_cmd.stdout);
        let stderr = String::from_utf8_lossy(&run_cmd.stderr);

        insta::assert_snapshot!(stdout);
        insta::assert_snapshot!(stderr);
        insta::assert_snapshot!(run_cmd.status);
    });
}

#[test]
fn typecheck_errors() {
    let std_package = Package::from_path("std", "std".into()).unwrap();
    let std_package = std_package.parse().unwrap();

    glob!("sources/typecheck_error/*", |file| {
        let main_package = Package::from_file(file.into()).unwrap();
        let main_package = main_package.parse().unwrap();

        let mut checker = Checker::new();
        let checked_std_package = checker.add_package(&std_package, &vec![]).unwrap();

        let err = checker
            .add_package(&main_package, &vec![(
                "std".into(),
                checked_std_package.package_id,
            )])
            .err()
            .unwrap();

        let mut buf = String::new();
        GraphicalReportHandler::new_themed(GraphicalTheme::none())
            .render_report(&mut buf, miette::Report::new(err).as_ref())
            .unwrap();

        insta::assert_snapshot!(buf);
    });
}

#[test]
fn parse_expr() {
    glob!("sources/parse_expr/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_expr(&source);

        insta::assert_debug_snapshot!(expr);
    });
}

#[test]
fn parse_stmt() {
    glob!("sources/parse_stmt/*", |file| {
        let source = fs::read_to_string(file).unwrap();
        let expr = crate::new_parser::Parser::parse_stmt(&source);

        insta::assert_debug_snapshot!(expr);
    });
}
