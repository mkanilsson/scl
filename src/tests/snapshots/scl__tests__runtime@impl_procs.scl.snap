---
source: src/tests/mod.rs
expression: code
input_file: src/tests/sources/runtime/impl_procs.scl
---
type :main.Vector = { w, w }
export function w $std.math.add(w %a, w %b) {
@start
	%.ss.0 =l alloc4 4
	%.ss.1 =l alloc4 4
	storew %a, %.ss.0
	storew %b, %.ss.1
@body
	%.sv.1 =w loadw %.ss.0
	%.sv.2 =w loadw %.ss.1
	%.binop.3 =w add %.sv.1, %.sv.2
	ret %.binop.3
}
export function w $main() {
@start
	%.ss.0 =l alloc4 8
	%.ss.1 =l alloc4 8
	%.ss.2 =l alloc4 4
	%.ss.3 =l alloc4 8
	%.ss.4 =l alloc4 8
	%.ss.5 =l alloc4 4
@body
	# Store value into Vector.x (offset: 0)
	%.offset.4 =l add %.ss.0, 0
	storew 5, %.offset.4
	# Store value into Vector.y (offset: 4)
	%.offset.5 =l add %.ss.0, 4
	storew 10, %.offset.5
	blit %.ss.0, %.ss.1, 8
	%.main..main.Vector.print.return_value.6 =w call $main..main.Vector.print(l %.ss.1)
	# Store value into Vector.x (offset: 0)
	%.offset.8 =l add %.ss.3, 0
	storew 7, %.offset.8
	# Store value into Vector.y (offset: 4)
	%.offset.9 =l add %.ss.3, 4
	storew 8, %.offset.9
	blit %.ss.3, %.ss.4, 8
	%.main..main.Vector.print.return_value.7 =w call $main..main.Vector.print(l %.ss.4)
	ret 0
}
export function $main..main.Vector.print(l %.this) {
@start
	%.ss.0 =l alloc4 4
	%.ss.1 =l alloc4 4
@body
	%.offset.12 =l add %.this, 0
	%.read.13 =w loadw %.offset.12
	%.offset.14 =l add %.this, 4
	%.read.15 =w loadw %.offset.14
	%.printf.return_value.10 =w call $printf(l $.str.11, ..., w %.read.13, w %.read.15)
	%.member_access.x.16 =l add %.this, 0
	storew 50, %.member_access.x.16
	%.offset.19 =l add %.this, 0
	%.read.20 =w loadw %.offset.19
	%.offset.21 =l add %.this, 4
	%.read.22 =w loadw %.offset.21
	%.printf.return_value.17 =w call $printf(l $.str.18, ..., w %.read.20, w %.read.22)
	ret
}
data $.str.11 = { b "Before: {x: %d, y: %d}\n", b 0 }
data $.str.18 = { b "After: {x: %d, y: %d}\n", b 0 }
